# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;














## Objetivo del proyecto ##

El presente proyecto tiene como objetivo analizar y proponer los libros y editoriales que mayores ventas han tenido en un plazo determinado, para sí promover aquellos más propensos a que sean adquiridos por nuevos clientes y a su vez, pueda servir de base para que otros autores o escritores dediquen sus esfuerzos en la creación de nuevas historias basándose en los gustos de los lectores, basándonos en reseñas y calificaciones otorgadas por diversos usuarios.

La obtención de estos datos los obtenderemos por medio de una base de datos descargada en línea utilizando el lenguaje SQL y exponiendo en cada paso del proyecto los resultados obtenidos. 

In [1]:


# import libraries
import pandas as pd
from sqlalchemy import create_engine

db_config = {
 'user': 'practicum_student', # username
 'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # password
 'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
 'port': 5432, # connection port
 'db': 'data-analyst-final-project-db' # the name of the database
 }
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'], db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})



<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>A resolver</b> ❗ - Muy bien logrado el paso de conexión a la base de datos, se nota que comprendes cómo usar `SQLAlchemy`. Sin embargo, es importante **no incluir datos sensibles como contraseñas o usuarios directamente en el código**.

Te recomiendo que uses **variables de entorno (`env`)** o archivos de configuración (como `.env`) para proteger esta información. Esto no solo mejora la seguridad, sino que también es una buena práctica profesional para proyectos colaborativos o de producción.

¡Sigue así! Este tipo de ajustes te hará escribir código más seguro y reutilizable. 💻🔐

</div>



In [2]:
# Apartado solo para ver todas las tablas disponibles para su descarga de la fuente
query = ''' SELECT *
FROM information_schema.tables 
WHERE table_schema = 'public';'''

data_raw = pd.io.sql.read_sql(query, con = engine)



In [3]:
print(data_raw.head)

<bound method NDFrame.head of                    table_catalog table_schema          table_name  table_type  \
0  data-analyst-final-project-db       public             ratings  BASE TABLE   
1  data-analyst-final-project-db       public  advertisment_costs  BASE TABLE   
2  data-analyst-final-project-db       public             authors  BASE TABLE   
3  data-analyst-final-project-db       public              orders  BASE TABLE   
4  data-analyst-final-project-db       public             reviews  BASE TABLE   
5  data-analyst-final-project-db       public              visits  BASE TABLE   
6  data-analyst-final-project-db       public               books  BASE TABLE   
7  data-analyst-final-project-db       public               users  BASE TABLE   
8  data-analyst-final-project-db       public          publishers  BASE TABLE   

  self_referencing_column_name reference_generation user_defined_type_catalog  \
0                         None                 None                      None 

In [4]:
# Comprobación de que se descargaron los archivos correctamente

query1 = ''' SELECT *
FROM books'''

books_df = pd.io.sql.read_sql(query1, con = engine)

query2 = ''' SELECT *
FROM authors'''

authors_df = pd.io.sql.read_sql(query2, con = engine)

query3 = ''' SELECT *
FROM publishers'''

publishers_df = pd.io.sql.read_sql(query3, con = engine)

query4 = ''' SELECT *
FROM ratings'''

ratings_df = pd.io.sql.read_sql(query4, con = engine)

query5 = ''' SELECT *
FROM reviews'''

reviews_df = pd.io.sql.read_sql(query5, con = engine)

print(books_df)
print(authors_df)
print(publishers_df)
print(ratings_df)
print(reviews_df)

     book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594 

### Número de libros publicados después del 1 de enero de 2000 ###

In [5]:
query = '''
SELECT 
    CAST(publication_date as date) as publication_date_converted,
    *
FROM
    books
'''

result = pd.io.sql.read_sql(query, con = engine)
print(result.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   publication_date_converted  1000 non-null   object
 1   book_id                     1000 non-null   int64 
 2   author_id                   1000 non-null   int64 
 3   title                       1000 non-null   object
 4   num_pages                   1000 non-null   int64 
 5   publication_date            1000 non-null   object
 6   publisher_id                1000 non-null   int64 
dtypes: int64(4), object(3)
memory usage: 54.8+ KB
None


In [6]:
query_count2 = '''SELECT
    MAX(CAST(publication_date AS date)) AS max_date,
    MIN(CAST(publication_date AS date)) AS min_date
    FROM
        books
'''

result_count2 = pd.io.sql.read_sql(query_count2, con = engine)
print(result_count2)

     max_date    min_date
0  2020-03-31  1952-12-01


In [7]:
query_count_ambos = '''
SELECT 
    COUNT(CASE WHEN CAST(publication_date as date) >= '2000-01-01' THEN 1 END) as libros_despues_2000,
    COUNT(CASE WHEN CAST(publication_date as date) < '2000-01-01' THEN 1 END) as libros_antes_2000
FROM
    books
'''

result_count_ambos = pd.io.sql.read_sql(query_count_ambos, con = engine)
print(result_count_ambos)

   libros_despues_2000  libros_antes_2000
0                  821                179


Vemos que la cantidad de libros publicados después del año 2000 hasta el 31 de marzo de 2020 fue mucho mayor que antes del año 2000, comprendiendo la base de datos desde el 1 de diciembre de 1952, a pesar de que es un plazo mucho menor. Esto puede deberse a que la tecnología no era tan avanzada como para publicar libros tan seguido o tal vez no había tantas editoriales que pudieran costear los costos de producción de los libros por ser la imprenta un método que no estaba tan avanzado ya a principios de los año 2000.

### Números de reseñas por usuario y calificación promedio por libro ###

In [8]:
query_count_review = '''
SELECT 
    COUNT(review_id) as cnt
FROM
    reviews
'''

result_count_reseñas = pd.io.sql.read_sql(query_count_review, con = engine)
print(result_count_reseñas)

    cnt
0  2793


In [9]:
query_count_ratings_avg = '''
SELECT 
    ratings.book_id AS book_id, 
    books.title AS title,
    books.publication_date AS publication_date,
    authors.author,
    AVG(ratings.rating) as avg_rating
FROM
    books
INNER JOIN ratings ON ratings.book_id=books.book_id
INNER JOIN authors ON authors.author_id=books.author_id
GROUP BY
    ratings.book_id,
    books.title,
    books.publication_date,
    authors.author
ORDER BY
    avg_rating DESC,
    title ASC

'''

result_avg_rating = pd.io.sql.read_sql(query_count_ratings_avg, con = engine)
print(result_avg_rating.head())

   book_id                                  title publication_date  \
0       57        Act of Treason (Mitch Rapp  #9)       2006-10-10   
1       17          A Dirty Job (Grim Reaper  #1)       2006-03-21   
2       20  A Fistful of Charms (The Hollows  #4)       2006-06-27   
3       62                          Alas  Babylon       2005-07-05   
4       76                            Angels Fall       2006-07-11   

              author  avg_rating  
0        Vince Flynn         5.0  
1  Christopher Moore         5.0  
2       Kim Harrison         5.0  
3          Pat Frank         5.0  
4       Nora Roberts         5.0  


Encontramos una cantidad de 2,793 reseñas de usuarios, así como la calificación promedio de cada libro, en la que una gran cantidad de libros de años recientes obtuvieron una calificación de 5 estrellas, la calificación perfecta. Esto puede deberse a que es más probable que los usuarios que que tienen acceso a dar calificaciones hayan leído libros más recientes que a los que son más antiguos.

### Editorial con mayor cantidad de libros publicados de más de 50 páginas ### 

In [10]:
query_top_publisher = '''
SELECT  
    publishers.publisher AS publisher,
    COUNT(publishers.publisher) as cnt
FROM
    books
INNER JOIN publishers ON publishers.publisher_id=books.publisher_id
WHERE 
    books.num_pages > 50
GROUP BY
    publisher
ORDER BY
    cnt DESC
'''

result_top_publisher = pd.io.sql.read_sql(query_top_publisher, con = engine)
print(result_top_publisher.head())

                  publisher  cnt
0             Penguin Books   42
1                   Vintage   31
2  Grand Central Publishing   25
3          Penguin Classics   24
4          Ballantine Books   19


La editorial 'Penguin Books' ha sido la editorial que más libros arriba de 50 páginas ha publicado, con 42 libros en total, seguido de la editorial 'Vintage' con 31 libros y en tercer lugar la editorial 'Grand Central Publishing' con 25 libros publicados. 

### Identifica el autor con más alta calificación ###

In [11]:
query_top_author = '''
SELECT  
    authors.author AS author,
    AVG(ratings.rating) AS avg_rating
FROM
    books
INNER JOIN ratings ON ratings.book_id=books.book_id
INNER JOIN authors ON authors.author_id=books.author_id
GROUP BY
    author
HAVING
    COUNT(ratings.rating) > 50
ORDER BY
    avg_rating DESC
    
'''

result_top_author = pd.io.sql.read_sql(query_top_author, con = engine)
print(result_top_author.head())

                              author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.288462
1                    Agatha Christie    4.283019
2  Markus Zusak/Cao Xuân Việt Khương    4.264151
3                     J.R.R. Tolkien    4.240964
4           Roald Dahl/Quentin Blake    4.209677


No es de extrañar que la autora de Harry Potter, J.K Rowling, sea la que mayor calificación promedio le otorgaron los usuarios, ya que por algo es de los libros más vendidos de todos los tiempos y con al menos más de 50 calificaciones de los lectores.

### Número promedio de reseñas de texto que calificaron más de 50 libros ###

In [12]:
query_avg_reviews = '''
SELECT  
    reviews.username AS username,
    COUNT(books.book_id) AS count
FROM
    books
INNER JOIN reviews ON reviews.book_id=books.book_id
GROUP BY
    username
HAVING
    COUNT(reviews.text) > 50
'''

result_avg_reviews = pd.io.sql.read_sql(query_avg_reviews, con = engine)
print(result_avg_reviews)

Empty DataFrame
Columns: [username, count]
Index: []


Para este ejercicio, no se encontraron usuarios que hayan dejado reseñas a más de 50 libros, por lo que sería buena idea incentivar a los lectores a dejar reseñas para así más usuarios nuevos tengan información para comprar más libros.


## Conclusión general ## 

Al término de las consultas en SQL para los libros, podemos aportar una sugerencia tanto para los nuevos autores como para que el equipo de marketing de cualquier tienda en línea pueda enfocar sus esfuerzos en promover ciertos libros con qué editorial.

La editorial con mayor cantidad de libros fue Penguin Books, la autora con mayor calificación fue J.K Rowling, autora de Harry Potter, y podemos observar también que los libros más recientes fueron los que mayor cantidad de reseñas y calificaciones obtuvieron, por lo que se sugiere no lanzar nuevamente libros que haya sido publicados hace mucho, o sea, no hacer reediciones ya que al público actual no se ve muy interesado en leer libros con mucho tiempo de publicación, así como enfocarse en temas de fantasía y misterios, como lo vimos con los autores como J.K Rowling, Agatha Christie y J.R.R. Tolkien, quiens han sido los autores mejor calificados y los resultados vistos en el cine hablan por sí mismos de sus libros. 

De igual manera, vimos que aún falta seguir promoviendo el hábito de la lectura, ya que ningún usuario ha escrito más de 50 reseñas. Esto igual podría deberse por la proliferación de las redes sociales, el uso de streaming, los videojuegos, que otorgan un entretenimiento más 'inmediato'. Esto puede resultar en que cada vez habrá menos usuarios que escriban tantas reseñas.



### # Comentario General del Revisor

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>


Quiero felicitarte por el esfuerzo que pusiste en este trabajo. Lograste conectarte correctamente a la base de datos, ejecutar múltiples consultas SQL relevantes, y presentar conclusiones claras que responden al objetivo del proyecto: entender qué libros y editoriales tienen mejor recepción entre los lectores. ¡Gran trabajo! 📚✨

---

### ✅ **Aspectos Positivos:**

* **Conexión a base de datos funcional**: Pudiste conectarte a PostgreSQL con SQLAlchemy y descargar todas las tablas correctamente.
* **Consultas SQL claras y bien estructuradas**: Usaste `JOIN`, `GROUP BY`, `HAVING`, y funciones agregadas de forma precisa para responder a las preguntas de negocio.
* **Enfoque estratégico**: El análisis muestra una buena comprensión del negocio editorial, considerando variables como número de páginas, calificaciones promedio y antigüedad del libro.
* **Conclusiones bien redactadas**: Interpretaste los resultados con criterio y lógica, proponiendo ideas accionables para marketing y nuevos autores.

---

### ⚠️ **Áreas de Mejora:**

* **Seguridad en el código**: Recuerda **no incluir directamente usuarios y contraseñas en el código**. Es mejor usar variables de entorno o un archivo `.env`. Esto es una práctica estándar en proyectos reales.
* **Conversión de fechas**: Aunque hiciste el casteo en SQL, también podrías trabajar las fechas con `pd.to_datetime()` para mayor flexibilidad en pandas.
* **Análisis estadístico adicional**: Te animamos a sumar visualizaciones más detalladas o métricas de dispersión para enriquecer los insights (por ejemplo, correlación entre páginas y ratings).

---


Nos vemos en el próximo proyecto.
— *Hesus García* ✌️😎

</div>


